ONSdigital/census-rh-service
# CR-126 RH Service endpoint to provide UAC and Case details

# Motivation and Context
[UAC and Case events](https://collaborate2.ons.gov.uk/confluence/pages/viewpage.action?spaceKey=CATD&title=Response+Management+Event+Dictionary) generated by RM will be stored as UAC and CaseCollection objects in Google Cloud Storage, this functionality is provided by [CR-118](https://github.com/ONSdigital/census-rh-service/pull/12). This data is required by the RH UI so this pull request implements an end point to provide the stored data. 

# What has changed
UniqueAccessCodeEndpoint added to retrieve UAC and Case data from storage via the UniqueAccessCodeService and return information required by RH UI to confirm an address for a UAC and send a claim to EQ to begin a questionnaire.

# How to test?
Unit tests provided for endpoint and underlying service.

You can also test manually when running RabbitMQ. You will need a Google Cloud Platform project setup and GCloud configured on your local machine. The environment variable GOOGLE_CLOUD_PROJECT should be set to the name of your project. Set the parameters in application.yml appropriately, for instance the port on which you are running RabbitMQ. When running the RH Service application go into the RabbitMQ management console. Publish the example message below to the Events exchange with the binding event.uac.updates:

{ 
     "event" : { 
           "type" : "UAC_UPDATED", 
           "source" : "CASE_SERVICE", 
           "channel" : "RM", 
           "dateTime" : "2011-08-12T20:17:46.384Z", 
           "transactionId" : "c45de4dc-3c3b-11e9-b210-d663bd873d93" 
      }, 
      "payload" : {
              "uac" : { 
                  "uacHash": "97a2e785f8cffc8f40b5e8fc626933bb8aedf3df82672299bf7aeafe68c99c93", 
                  "active" : true, 
                  "questionnaireId":"1110000009", 
                  "caseType":"H", 
                  "region":"E", 
                  "caseId":"dc4477d1-dd3f-4c69-b181-7ff725dc9fa4", 
                  "collectionExerciseId":"a66de4dc-3c3b-11e9-b210-d663bd873d93" 
              } 
      } 
}
An object should appear in the Google Cloud Storage Bucket <project-name>-uac.

Publish the following example message to the Events exchange with the binding event.case.lifecycle:

{ 
     "event" : { 
           "type" : "CASE_UPDATED", 
           "source" : "CASE_SERVICE", 
           "channel" : "RM", 
           "dateTime" : "2011-08-12T20:17:46.384Z", 
           "transactionId" : "c45de4dc-3c3b-11e9-b210-d663bd873d93" 
     }, 
     "payload" : { 
             "collectionCase" : { 
                        "id":"dc4477d1-dd3f-4c69-b181-7ff725dc9fa4", 
                        "caseRef": "10000000010", 
                        "survey": "CENSUS", 
                        "collectionExerciseId": "a66de4dc-3c3b-11e9-b210-d663bd873d93", 
                        "address": { 
                                "addressLine1": "1 main street", 
                                "addressLine2": "upper upperingham", 
                                "addressLine3": "", 
                                "townName": "upton", 
                                "postcode": "UP103UP", 
                                "region": "E", 
                                "latitude": "50.863849", 
                                "longitude": "-1.229710", 
                                "uprn":"123456", 
                                "arid":"XXXXX", 
                                "addressType":"CE", 
                                "estabType":"XXX" 
                        }, 
                        "contact": { 
                                "title":"Ms", 
                                "forename":"jo", 
                                "surname":"smith", 
                                "email":"me@example.com", 
                                "telNo":"+447890000000" 
                        }, 
                        "state":"ACTIONABLE", 
                        "actionableFrom":"2011-08-12T20:17:46.384Z" 
             } 
     } 
}  
An object should appear in the Google Cloud Storage Bucket <project-name>-case. The system now has dummy data in GCS ready to exercise the endpoint. 

Using Postman or Curl with the appropriate authorisation header issue the GET request:
       http://localhost:8171/uacs/w4nwwpphjjpt
The value "w4nwwpphjjpt" is the UAC value which is SHA-256 hashed as uacHash in the UAC event message and these must correlate. You should receive a response such as:

{
    "uac": "w4nwwpphjjpt",
    "active": true,
    "questionnaireId": 1110000009,
    "caseType": "H",
    "region": "E",
    "caseId": "dc4477d1-dd3f-4c69-b181-7ff725dc9fa4",
    "collectionExerciseId": "a66de4dc-3c3b-11e9-b210-d663bd873d93",
    "address": {
        "addressLine1": "1 main street",
        "addressLine2": "upper upperingham",
        "addressLine3": "",
        "townName": "upton",
        "postcode": "UP103UP",
        "uprn": "123456"
    }
}

Obviously by changing the objects stored in GCS, no UAC object, no Case object, you can simulate the various scenarios for testing.
